### PR TITLE
config: Deprecate opaque Struct config, again.

### LINKED ISSUE #8403: re-deprecate opaque Struct config
Structs are deprecated once and un-deprecated in #6346. Now protobuf hashing are sorted and moving away from hashing in long term (#8231, #8301). So we can deprecate this and migrate all docs/examples.

cc @htuch @alyssawilk @moderation

### PR DESCRIPTION
**Pull request by yanavlasov**
Deprecate opaque Struct config, again.
Fix unit tests to use the google.protobuf.Any

Reverts PR #6346
Fixes #8403

Risk Level: Low
Testing: Unit Tests
Docs Changes: N/A
Release Notes: deprecation notice

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

**Comment by yanavlasov** [comment]
@lizan , @htuch please take a look at the initial set of changes. I still need to fix two unit tests with --define deprecated_features=disabled and one without.
Let me know if this is moving in the right direction.

**Comment by lizan** [comment]
Thanks @yanavlasov, this looks great and is definitely in right direction! You also need to run the proto format check fix to reflect in changes v3alpha and update deprecated doc.

File: docs/root/intro/deprecated.rst [hunk 1]
```diff
@@ -39,6 +39,8 @@ Version 1.12.0 (pending)
   has been deprecated in favor of the `traffic_direction` field in
   :ref:`Listener <envoy_api_msg_Listener>`. The latter takes priority if
   specified.
+* Use of google.protobuf.Struct for extension opaque configs is deprecated. Use google.protobuf.Any instead or pack
+  udpa.type.v1.TypedStruct in google.protobuf.Any.
```
Comment (htuch) [inline docs/root/intro/deprecated.rst]: Are you planning on implementing this in this PR? We should probably have the new solution implemented before deprecating :) Otherwise, yeah, we're heading in the right direction here.

Comment (lizan) [reply docs/root/intro/deprecated.rst]: Let's do that in a separate PR and merge that one before this.

**Comment by yanavlasov** [comment]
All tests pass (for now). I've updated docs/root/intro/deprecated.rst Let me know if there is anything else that needs to be done in this PR.
I will work on parsing TypedStruct in config protos next.

File: test/common/router/config_impl_test.cc [hunk 2]
```diff
@@ -6088,11 +6085,10 @@ name: foo
           seconds: 123
 )EOF";
 
-    checkNoPerFilterConfig(yaml);
-  }
+  checkNoPerFilterConfig(yaml);
 }
 
-TEST_F(PerFilterConfigsTest, RouteLocalConfig) {
+TEST_F(PerFilterConfigsTest, DEPRECATED_FEATURE_TEST(RouteLocalConfig)) {
```
Comment (htuch) [inline test/common/router/config_impl_test.cc]: Why are only some of the tests upgraded, while others are left behind. Do we want these tests to be deleted in v3?

File: test/integration/integration.cc [hunk 3]
```diff
@@ -432,10 +432,13 @@ void BaseIntegrationTest::createGeneratedApiTestServer(const std::string& bootst
...
     const char* rejected = "listener_manager.lds.update_rejected";
+    const char* failed = "listener_manager.lds.update_failure";
     while ((test_server_->counter(success) == nullptr ||
             test_server_->counter(success)->value() == 0) &&
-           (!allow_lds_rejection || test_server_->counter(rejected) == nullptr ||
-            test_server_->counter(rejected)->value() == 0)) {
+           (!allow_lds_rejection || ((test_server_->counter(rejected) == nullptr ||
+                                      test_server_->counter(rejected)->value() == 0) &&
+                                     (test_server_->counter(failed) == nullptr ||
+                                      test_server_->counter(failed)->value() == 0)))) {
```
Comment (htuch) [inline test/integration/integration.cc]: Why were these changes needed?

File: test/common/router/config_impl_test.cc (hunk 2 above)
Comment (yanavlasov) [reply test/common/router/config_impl_test.cc]: All of test methods, except this one (RouteLocalConfig) tested both Any and Struct based configurations in the same test method. I have split these test methods into two testing Any and Struct separately and deprecating the Struct tests. The RouteLocalConfig was already split into separate methods.
The TypedConfigFilterError test above tests error condition when both Struct and Any are present in the config and thus deprecated as well, as the Struct field is removed in v3.

File: docs/root/intro/deprecated.rst (hunk 1 above)
Comment (yanavlasov) [reply docs/root/intro/deprecated.rst]: PR #8515

File: test/extensions/access_loggers/grpc/tcp_grpc_access_log_integration_test.cc [hunk 4]
```diff
@@ -53,10 +53,11 @@ class TcpGrpcAccessLogIntegrationTest : public Grpc::GrpcClientIntegrationParamT
     config_helper_.addConfigModifier([this](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
       auto* listener = bootstrap.mutable_static_resources()->mutable_listeners(0);
       auto* filter_chain = listener->mutable_filter_chains(0);
-      auto* config_blob = filter_chain->mutable_filters(0)->mutable_config();
+      auto* config_blob = filter_chain->mutable_filters(0)->mutable_typed_config();
 
+      ASSERT_TRUE(config_blob->Is<envoy::config::filter::network::tcp_proxy::v2::TcpProxy>());
       envoy::config::filter::network::tcp_proxy::v2::TcpProxy tcp_proxy_config;
-      TestUtility::jsonConvert(*config_blob, tcp_proxy_config);
+      config_blob->UnpackTo(&tcp_proxy_config);
```
Comment (htuch) [inline test/extensions/access_loggers/grpc/tcp_grpc_access_log_integration_test.cc]: You probably want to use https://github.com/envoyproxy/envoy/blob/master/source/common/protobuf/utility.h#L290 instead of `UnpackTo`, since this verifies the result.

File: test/integration/integration.cc [hunk 5]
```diff
@@ -431,11 +431,20 @@ void BaseIntegrationTest::createGeneratedApiTestServer(const std::string& bootst
     // needs to know about the bound listener ports.
     auto end_time = time_system_.monotonicTime() + TestUtility::DefaultTimeout;
     const char* success = "listener_manager.listener_create_success";
+    // Loading invalid configuration may result in different error counters to be incremented.
+    // Loading invalid configuration using the google.protobuf.Any results in the
+    // "listener_manager.lds.update_failure" incremented. Loading invalid configuration using
+    // google.protobuf.Struct (deprecated in API v2) results in the
+    // "listener_manager.lds.update_rejected" incremented. See FilesystemSubscriptionImpl::refresh()
```
Comment (htuch) [inline test/integration/integration.cc]: This seems kind of wrong though, is it reasonable to easily change the behavior of typed config to be consistent with Struct? AFAICT it should be a rejection rather than a failure ideally.

File: test/integration/dynamic_validation_integration_test.cc [hunk 6]
```diff
@@ -117,7 +117,8 @@ TEST_P(DynamicValidationIntegrationTest, LdsFilterRejected) {
   if (reject_unknown_dynamic_fields_) {
     EXPECT_EQ(0, test_server_->counter("listener_manager.lds.update_success")->value());
     // LDS API parsing will reject due to unknown HCM field.
-    EXPECT_EQ(1, test_server_->counter("listener_manager.lds.update_rejected")->value());
+    EXPECT_TRUE(1 == test_server_->counter("listener_manager.lds.update_rejected")->value() ||
```
Comment (htuch) [inline test/integration/dynamic_validation_integration_test.cc]: Is this the same case as `integration.cc` below?

Comment (yanavlasov) [reply test/integration/dynamic_validation_integration_test.cc]: Done

File: test/integration/integration.cc (hunk 5 above)
Comment (yanavlasov) [reply test/integration/integration.cc]: I've changed the exception type that is thrown when unknown fields are detected. So the update_rejected counter is incremented for the typed configs as well.

File: test/extensions/access_loggers/grpc/tcp_grpc_access_log_integration_test.cc (hunk 4 above)
Comment (yanavlasov) [reply test/extensions/access_loggers/grpc/tcp_grpc_access_log_integration_test.cc]: Done
